RemoteDOMWindow::close needs same security checks as LocalDOMWindow::close

achristensen07 opened the pull request:
#### 58b0fdf84a3249ec5964ceb9fddf236f2c99b01c
<pre>
RemoteDOMWindow::close needs same security checks as LocalDOMWindow::close
<a href="https://bugs.webkit.org/show_bug.cgi?id=268530">https://bugs.webkit.org/show_bug.cgi?id=268530</a>
<a href="https://rdar.apple.com/117381050">rdar://117381050</a>

Reviewed by Charlie Wolfe.

I moved the checks to DOMWindow and shared them.

I added RemoteDOMWindow::frameDetached so that frame.remove() followed immediately by
frame.contentWindow.closed returns true with a RemoteDOMWindow like it does a LocalDOMWindow
because the frame gets detached.

WebPageProxy::closeRemoteFrame isn&apos;t really needed any more because we already have
checks that the close() call is only coming from a main frame, so it is closing the page.
Just call WebPage::closePage to close the page without having an unneeded code path.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::close):
(WebCore::DOMWindow::console const):
(WebCore::DOMWindow::checkedConsole const):
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/FrameDestructionObserver.cpp:
(WebCore::FrameDestructionObserver::~FrameDestructionObserver):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::closePage):
(WebCore::LocalDOMWindow::console const): Deleted.
(WebCore::LocalDOMWindow::checkedConsole const): Deleted.
(WebCore::LocalDOMWindow::close): Deleted.
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::closePage):
(WebCore::RemoteDOMWindow::frameDetached):
(WebCore::RemoteDOMWindow::close): Deleted.
* Source/WebCore/page/RemoteDOMWindow.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::frameDetached):
* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::closeRemoteFrame): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::closePage):
(WebKit::WebRemoteFrameClient::close): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:

Canonical link: <a href="https://commits.webkit.org/273911@main">https://commits.webkit.org/273911@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/13cfe0f55e4f1de8bca6d507b6a1a18c06b39690

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37171 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/16072 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39467 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39743 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/33174 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38464 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18650 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13215 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/31668 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37734 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13569 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/32722 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11812 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/11814 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/33355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/41000 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33664 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34382 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/37722 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12142 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9936 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35866 "Passed tests") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/13764 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12498 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4812 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/13028 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

charliewolfe commented on Source/WebCore/page/LocalDOMWindow.h:
```diff
@@ -257,9 +255,6 @@ class LocalDOMWindow final
     RefPtr<WebKitPoint> webkitConvertPointFromNodeToPage(Node*, const WebKitPoint*) const;
 
-    PageConsoleClient* console() const;
-    CheckedPtr<PageConsoleClient> checkedConsole() const;
```
The `PageConsoleClient` forward declaration can be removed from this file.

charliewolfe commented on Source/WebCore/page/LocalDOMWindow.cpp:
```diff
@@ -753,20 +753,6 @@ BarProp& LocalDOMWindow::toolbar()
-{
-    // FIXME: This should not return nullptr when frameless.
-    if (!isCurrentlyDisplayedInFrame())
-        return nullptr;
```
Doesn't look like this part was moved, is it ok to drop this?

achristensen07 replied:
Yes.

achristensen07 committed:
RemoteDOMWindow::close needs same security checks as LocalDOMWindow::close
https://bugs.webkit.org/show_bug.cgi?id=268530
rdar://117381050

Reviewed by Charlie Wolfe.

I moved the checks to DOMWindow and shared them.

I added RemoteDOMWindow::frameDetached so that frame.remove() followed immediately by
frame.contentWindow.closed returns true with a RemoteDOMWindow like it does a LocalDOMWindow
because the frame gets detached.

WebPageProxy::closeRemoteFrame isn't really needed any more because we already have
checks that the close() call is only coming from a main frame, so it is closing the page.
Just call WebPage::closePage to close the page without having an unneeded code path.

* LayoutTests/platform/mac-site-isolation/TestExpectations:
* Source/WebCore/page/DOMWindow.cpp:
(WebCore::DOMWindow::close):
(WebCore::DOMWindow::console const):
(WebCore::DOMWindow::checkedConsole const):
* Source/WebCore/page/DOMWindow.h:
* Source/WebCore/page/FrameDestructionObserver.cpp:
(WebCore::FrameDestructionObserver::~FrameDestructionObserver):
* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::closePage):
(WebCore::LocalDOMWindow::console const): Deleted.
(WebCore::LocalDOMWindow::checkedConsole const): Deleted.
(WebCore::LocalDOMWindow::close): Deleted.
* Source/WebCore/page/LocalDOMWindow.h:
* Source/WebCore/page/RemoteDOMWindow.cpp:
(WebCore::RemoteDOMWindow::closePage):
(WebCore::RemoteDOMWindow::frameDetached):
(WebCore::RemoteDOMWindow::close): Deleted.
* Source/WebCore/page/RemoteDOMWindow.h:
* Source/WebCore/page/RemoteFrame.cpp:
(WebCore::RemoteFrame::frameDetached):
* Source/WebCore/page/RemoteFrameClient.h:
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::closeRemoteFrame): Deleted.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.cpp:
(WebKit::WebRemoteFrameClient::closePage):
(WebKit::WebRemoteFrameClient::close): Deleted.
* Source/WebKit/WebProcess/WebCoreSupport/WebRemoteFrameClient.h:

Canonical link: https://commits.webkit.org/273911@main
